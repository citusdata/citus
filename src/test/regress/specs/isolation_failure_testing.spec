# see that we can cause a session to block at a point of our choosing

session "s1"

setup { BEGIN; SET deadlock_timeout TO '100ms'; }
step "s1-lock-1" { SELECT pg_advisory_xact_lock(1); }
step "s1-lock-2" { SELECT pg_advisory_xact_lock(2); }
step "s1-commit" { COMMIT; }

session "s2"

setup { BEGIN; SET deadlock_timeout TO '100ms'; SET application_name TO 'gdb-process' }
step "s2-lock-1" { SELECT pg_advisory_xact_lock(1); }
step "s2-lock-2" { SELECT pg_advisory_xact_lock(2); }
step "s2-commit" { COMMIT; }

step "s2-lock-1-gdb"
{
  SELECT 1;
}

session "gdb"

step "gdb-break"
{
  WITH pid AS (
    SELECT pid FROM pg_stat_activity WHERE application_name = 'gdb-process'
  ) SELECT citus.gdb_attach(pid.pid) FROM pid;

  SELECT citus.run_command('!break NeedsDistributedPlanning');
  SELECT citus.run_command('continue');
}

# a regular deadlock is detected
permutation "s2-lock-2" "s1-lock-1" "s1-lock-2" "s2-lock-1" "s2-commit" "s1-commit"
# now try to inject a deadlock with gdb
permutation "s2-lock-2" "s1-lock-1" "s1-lock-2" "gdb-break" "s2-lock-1-gdb" "s2-commit" "s1-commit"
