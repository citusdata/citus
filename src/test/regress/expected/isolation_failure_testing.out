Parsed test spec with 3 sessions

starting permutation: s2-lock-2 s1-lock-1 s1-lock-2 s2-lock-1 s2-commit s1-commit
step s2-lock-2: SELECT pg_advisory_xact_lock(2);
pg_advisory_xact_lock

               
step s1-lock-1: SELECT pg_advisory_xact_lock(1);
pg_advisory_xact_lock

               
step s1-lock-2: SELECT pg_advisory_xact_lock(2); <waiting ...>
step s2-lock-1: SELECT pg_advisory_xact_lock(1); <waiting ...>
step s2-lock-1: <... completed>
pg_advisory_xact_lock

               
step s1-lock-2: <... completed>
error in steps s2-lock-1 s1-lock-2: ERROR:  deadlock detected
step s2-commit: COMMIT;
step s1-commit: COMMIT;

starting permutation: s2-lock-2 s1-lock-1 s1-lock-2 gdb-break s2-lock-1-gdb s2-commit s1-commit
step s2-lock-2: SELECT pg_advisory_xact_lock(2);
pg_advisory_xact_lock

               
step s1-lock-1: SELECT pg_advisory_xact_lock(1);
pg_advisory_xact_lock

               
step s1-lock-2: SELECT pg_advisory_xact_lock(2); <waiting ...>
step gdb-break: 
  WITH pid AS (
    SELECT pid FROM pg_stat_activity WHERE application_name = 'gdb-process'
  ) SELECT citus.gdb_attach(pid.pid) FROM pid;

  SELECT citus.run_command('!break NeedsDistributedPlanning');
  SELECT citus.run_command('continue');

gdb_attach     

               
run_command    

               
run_command    

               
step s2-lock-1-gdb: 
  SELECT 1;
 <waiting ...>
step s2-lock-1-gdb: <... completed>
step s1-lock-2: <... completed>
pg_advisory_xact_lock

               
error in steps s2-lock-1-gdb s1-lock-2: ERROR:  deadlock detected
step s2-commit: COMMIT;
step s1-commit: COMMIT;
